app/controllers/candidate_interface/sign_in_controller.rb

```diff
@@ -80,9 +75,38 @@ def authenticate
 
     def expired
       raise unless FeatureFlag.active?('improved_expired_token_flow')
-      return redirect_to candidate_interface_sign_in_path unless params[:id]
 
-      @candidate = Candidate.find(params[:id])
+      if params[:u].blank?
+        redirect_to candidate_interface_sign_in_path
+        return
+      end
+
+      begin
+        candidate_id = Encryptor.decrypt(params[:u])
+        @candidate = Candidate.find(candidate_id)
+      rescue ActiveSupport::MessageEncryptor::InvalidMessage
+        redirect_to candidate_interface_sign_in_path
+      end
+    end
+
+    def create_from_expired_token
```

Comment:
I think this can be simplified a bit:

```rb
def create_from_expired_token
  render_404 unless FeatureFlag.active?('improved_expired_token_flow')

  candidate_id = Encryptor.decrypt(params.fetch(:u))
  candidate = Candidate.find(candidate_id)
  MagicLinkSignIn.call(candidate: candidate)
  add_identity_to_log candidate.id
  redirect_to candidate_interface_check_email_sign_in_path
rescue ActiveSupport::MessageEncryptor::InvalidMessage
  render 'errors/not_found', status: :forbidden
end
```

- Rely on `fetch` to check for token so that we display a BadRequest error page if it's not there
- Render a forbidden status code when the token is incorrect
- Rely on `Candidate.find` to raise an exception if the user id isn't found
- We don't need the instance variable anymore